src/Database/PostgreSQL/PQTypes/DerivingVia.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE AllowAmbiguousTypes, TypeApplications #-}
+module Database.PostgreSQL.PQTypes.DerivingVia (
+  -- | Helpers, to be used with `deriving via`.
+    SQLEnum(..)
+  , SQLEnumEncoding(..)
+  , SQLEnumAsText(..)
+  , SQLEnumAsTextEncoding(..)
+  , isInjective  -- ^ for doctests
+  ) where
+
+import Control.Exception (SomeException(..), throwIO)
+import Data.List.Extra (enumerate, nubSort)
+import Data.Map.Strict (Map)
+import Data.Text (Text)
+import Data.Typeable
+import Database.PostgreSQL.PQTypes
+import qualified Data.Map.Strict as Map
+
+-- | To be used in doctests to prove injectivity of encoding functions.
+--
+-- >>> isInjective (id :: Bool -> Bool)
+-- True
+--
+-- >>> isInjective (\(_ :: Bool) -> False)
+-- False
+isInjective :: (Enum a, Bounded a, Eq a, Eq b) => (a -> b) -> Bool
+isInjective f = null [ (a, b) | a <- enumerate, b <- enumerate, a /= b, f a == f b ]
+
+-- | Internal helper; given a list of values, decompose it into a list of
+-- intervals.
+-- >>> intervals [42,2,1,0,3,88,-1,43,42]
+-- [(-1,3),(42,43),(88,88)]
+--
+-- prop> nubSort xs == concatMap (\(l,r) -> [l .. r]) (intervals xs)
+intervals :: forall  a . (Enum a, Ord a) => [a] -> [(a, a)]
```

Comment:
This can be put into a where clause of the appropriate `FromSQL` instance.

Reply:
Yup, needs to be a global definition to be able to doctest it...